File: server/app/models/stock_trade.rb
# == Schema Information
#
# Table name: stock_trades
#
#  id               :uuid             not null, primary key
#  stock_holding_id :uuid             not null
#  quantity         :integer          not null
#  enter_price      :decimal(15, 2)
#  exit_price       :decimal(15, 2)
#  exit_time        :datetime
#  created_at       :datetime         not null
#  updated_at       :datetime         not null
#

class StockTrade < ApplicationRecord
  belongs_to :stock_holding
  has_one :stock, through: :stock_holding

  validates :stock_holding_id, presence: true

  alias holding stock_holding

  after_create_commit { StockTradeBroadcastJob.perform_later self }

  def stock_name
    stock.name
  end
end

File: phoenix_api/web/models/stock_trade.ex
defmodule PhoenixApi.StockTrade do
  use PhoenixApi.Web, :model

  schema "stock_trades" do
    belongs_to :stock_holding, PhoenixApi.StockHolding
    field :quantity, :integer
    field :enter_price, :decimal
    field :exit_price, :decimal
    field :exit_time, Ecto.DateTime

    timestamps()
  end

  @doc """
  Builds a changeset based on the `struct` and `params`.
  """
  def changeset(struct, params \\ %{}) do
    struct
    |> cast(params, [:stock_holding_id, :quantity, :enter_price, :exit_price, :exit_time])
    |> validate_required([:stock_holding_id, :quantity, :enter_price, :exit_price, :exit_time])
  end
end
